packages/vite/src/node/build.ts

```diff
@@ -501,7 +501,8 @@ async function doBuild(
     ssr ? config.plugins.map((p) => injectSsrFlagToHooks(p)) : config.plugins
   ) as Plugin[]
 
-  const userExternal = options.rollupOptions?.external
+  const userExternal =
+    (ssr && config.ssr?.external) || options.rollupOptions?.external
```

Comment:
Isn't `ssr.external` only for modules with state in ssr?

[playground/ssr-deps/server.js#L44](https://github.com/vitejs/vite/blob/d76db0cae645beaecd970d95b4819158c5dd568a/playground/ssr-deps/server.js#L44)